src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,162 @@
+            }
+            return array;
+        }
+        throw new RuntimeException("Given object cannot be converted to an array");
+    }
 
+    private void fastRemove(Object[] es, int i) {
+        final int newSize;
+        if ((newSize = size - 1) > i) {
+            System.arraycopy(es, i + 1, es, i, newSize - i);
+        }
+        es[size = newSize] = null;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index is incorrect " + index);
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        rangeCheckForAdd(index);
+        final int s;
+        Object[] elementData;
+        if ((s = size) == (elementData = this.elementData).length) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                s - index);
+        elementData[index] = value;
+        size = s + 1;
+    }
+
+    @Override
+    public void add(T value) {
+        add(value, elementData, size);
+    }
 
+    private void add(T value, Object[] elementData, int s) {
+        if (s == elementData.length) {
+            elementData = grow();
+        }
+        elementData[s] = value;
+        size = s + 1;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] array = toArray(list);
+        int numNew = array.length;
+        Object[] elementData;
+        final int s;
+        if (numNew > (elementData = this.elementData).length - (s = size)) {
+            elementData = grow(s + numNew);
+        }
+        System.arraycopy(array, 0, elementData, s, numNew);
+        size = s + numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        final Object[] es = elementData;
+
+        T oldValue = (T) es[index];
+        fastRemove(es, index);
+
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        final Object[] es = elementData;
+        final int size = this.size;
+        int i = 0;
+        boolean found = false;
+
+        if (element == null) {
+            for (; i < size; i++) {
+                if (es[i] == null) {
+                    found = true;
+                    break;
+                }
+            }
+        } else {
+            for (; i < size; i++) {
+                if (element.equals(es[i])) {
+                    found = true;
+                    break;
+                }
+            }
+        }
+        if (!found) {
+            throw new NoSuchElementException("Element not found");
+        }
```

Comment:
Ensure that all exception messages are consistent and informative. The `remove(T element)` method throws a `NoSuchElementException` with a different message style compared to other methods.